### DISCUSSION TITLE
Which version of typescript we should use?

### DISCUSSION THREAD

**Discussion by tomtobac** [twilio-labs/paste#1120] [Q&A] [answered]
Hey folks, I recently updated the dependencies of `@twilio-paste` to use only `core` and `icons` (see screenshot)  but now our components doesn't have styles and if we keep a lower typescript dependency `3.7.5` we are getting this weird errors:

Updated dependencies:
![image](https://user-images.githubusercontent.com/10562610/106138882-fc2b7200-616c-11eb-83ac-daed7432a9ea.png)

E.g.: Using typescript `3.7.5`

Import:
```ts
import { Box, Button, Stack } from '@twilio-paste/core';
```

Error: 
![image](https://user-images.githubusercontent.com/10562610/106138915-0a798e00-616d-11eb-97ce-4fb8b5c8c420.png)

![image](https://user-images.githubusercontent.com/10562610/106138747-d3a37800-616c-11eb-8ab8-7e44877645b4.png)

If we upgrade it to typescript `4.1.3` the errors of typescript disappear but the components doesn't have style at all :(

![image](https://user-images.githubusercontent.com/10562610/106139233-78be5080-616d-11eb-9fba-54fea2f7117e.png)

Only form components, modals and button secondary. Any ideas?

My pr with the changes and repo: https://code.hq.twilio.com/voice-n-video-console/vnv-console-video/pull/100/files

**Comment by SiTaggart** [comment]
The version of typecript shouldn't matter. This looks similar to a previous issue that got raised about weird type issues https://github.com/twilio-labs/paste/issues/827.

It turned out to be a version of `@types/react`, which we have tried to pin as a dependency so you install the correct one. Problem is depending on your own dep tree, you might not be using the most current one.

try `yarn why @types/react` and see what version its resolving too. You might have to add a resolution to your package json file if it's not getting the right one.

See if that fixes the type errors first.

I suspect the styles are caused by a different issues

**Reply by andioneto** [reply]
@tomtobac I think for npm the command is `npm ls @types/react`. Also, following this thread to see how to resolve. Thanks for raising this!

**Reply by tomtobac** [reply]
I ran `npm ls @types/react`:
```
vnv-console-video@1.0.0 /Users/tcabotvallori/Projects/vnv-console-video
├─┬ @testing-library/react-hooks@3.3.0
│ └─┬ @types/testing-library__react-hooks@3.4.1
│   └─┬ @types/react-test-renderer@17.0.0
│     └── @types/react@16.9.35  deduped
├─┬ UNMET PEER DEPENDENCY @twilio-paste/core@4.3.1
│ └─┬ @twilio-paste/types@3.1.1
│   ├── @types/react@16.14.2
│   └─┬ @types/react-dom@16.9.10
│     └── @types/react@16.14.2  deduped
├─┬ @types/enzyme@3.10.4
│ └── @types/react@16.9.35  deduped
├── @types/react@16.9.35
├─┬ @types/react-datepicker@3.1.3
│ └── @types/react@16.9.35  deduped
├─┬ @types/react-dom@16.9.5
│ └── @types/react@16.9.35  deduped
├─┬ @types/react-router@5.1.4
│ └── @types/react@16.9.35  deduped
└─┬ @types/react-router-dom@5.1.3
  └── @types/react@16.9.35  deduped

npm ERR! peer dep missing: @twilio-paste/core@4.3.0, required by @twiliointernal/vnv-client-toolbox@1.1.1
````

And my current versions as devDependencies:
```
"@types/react": "16.9.35",
"@types/react-dom": "16.9.5",
```

**Reply by tomtobac** [reply]
@SiTaggart if typescript version doesn't matter, why if I use the `4.1.3` I dont have this errors? 

This is my root component:
```tsx
import React, { lazy, Suspense } from 'react';
import { Theme } from '@twilio-paste/core/theme';

import { NotificationStoreProvider } from 'common/context/notifications';
import AppLoader from 'common/components/AppLoader';

const Settings = lazy(() =>
	import(/* webpackChunkName: "Settings" */ './components/RecordingSettingsV2'),
);

export const RecordingSettings = () => {
	return (
		<Theme.Provider theme="console">
			<NotificationStoreProvider>
				<Suspense fallback={<AppLoader />}>
					<Settings />
				</Suspense>
			</NotificationStoreProvider>
		</Theme.Provider>
	);
};
```

**Reply by TheSisb** [reply]
@tomtobac Any chance you can provide a repro on Codesandbox?  I tried myself but couldn't reproduce your particular issue: https://codesandbox.io/s/sample-using-core-and-icons-1qcq7. Maybe there's something going on with your `tsconfig.json` or `package-lock.json`?

Unrelated, but I also noted you're importing like this (from your message):
```js
import { Box, Button, Stack } from '@twilio-paste/core';
```
We now recommend importing like this:
```js
import { Box } from '@twilio-paste/core/box';
import { Button } from '@twilio-paste/core/button';
import { Stack } from '@twilio-paste/core/stack';
```

For an explanation, please see https://paste.twilio.design/getting-started/engineering#how-the-core-package-stays-lean
We have a codemod to help with migration: https://paste.twilio.design/libraries/codemods

**Reply by tomtobac** [reply]
hey @TheSisb , I can not reproduce in codesanbox, my guess it's something related with how we parse with ts-loader and our version of typescript..  I can share our `tsconfig.json`

- tsconfig.json
```json
{
	"compilerOptions": {
		"target": "ES2019",
		"lib": ["dom", "dom.iterable", "esnext"],
		"allowJs": true,
		"checkJs": false,
		"esModuleInterop": true,
		"allowSyntheticDefaultImports": true,
		"strict": true,
		"forceConsistentCasingInFileNames": true,
		"suppressImplicitAnyIndexErrors": true,
		"module": "esnext",
		"moduleResolution": "node",
		"resolveJsonModule": true,
		"isolatedModules": true,
		"noEmit": false,
		"experimentalDecorators": true,
		"jsx": "react",
		"skipLibCheck": true,
		"baseUrl": ".",
		"outDir": "tsout.js",
		"paths": {
			"console-core-ui/utils/*": ["node_modules/@console/vnv-console-core-ui/src/common/utils/*"],
			"console-core-ui/console/*": [
				"node_modules/@console/vnv-console-core-ui/src/common/console/*"
			],
			"common/components/*": ["client/src/common/components/*"],
			"common/context/*": ["client/src/common/context/*"],
			"common/enums": ["client/src/common/enums/index.ts"],
			"common/hooks/*": ["client/src/common/hooks/*"],
			"common/strings/*": ["client/src/common/strings/*"],
			"common/strings": ["client/src/common/strings/index.ts"],
			"common/styles": ["client/src/common/styles/index.ts"],
			"common/types/*": ["client/src/common/types/*"],
			"common/utils": ["client/src/common/utils/index.ts"]
		}
	},
	"include": ["client/src/**/*"]
}
```

Plus, using Typescript version `3.9.5` and `@twilio-paste/core@3.4.0` and it worked almost everything but the modals doesn't work well.. As you can see, the tokens are not parsed

![image](https://user-images.githubusercontent.com/10562610/106197773-551df900-61b3-11eb-9cac-606d1177d7fb.png)

p.d.: thanks for the heads up for the new imports, I ran the codemod and it worked, very cool 👌

**Reply by corinne-nagel** [reply]
Following up here, @tomtobac did you find a solution to this typescript issue?

**Reply by synnottd** [reply]
@tomtobac I think `"@types/react": "16.9.35",` is your problem, when we had that issue I am pretty sure yarn was defaulting to that rather than a newer one. I'd suggest trying to update to at least `16.9.44` that's the version that worked for us.

**Reply by serifluous** [reply]
Hey @tomtobac ! ^ @synnottd's on the right track here, though the version we require is actually `16.9.53`. Let us know if that helps you out.

**Reply by kate-porter** [reply]
Hey @tomtobac - just wanted to followup on this and make sure updating your version solved this for ya? Let us know!

**Reply by tomtobac** [reply]
hey @katieporter , at the end we pinned this version and it worked
```
"@twilio-paste/core": "4.1.0",
"@twilio-paste/icons": "3.12.1",
 "typescript": "3.9.4",
```
I will try with the `"@types/react": "16.9.35"` version whenever I have a chance and I will let you know, thanks a lot!

**Reply by kate-porter** [reply]
Great! Glad that worked for ya.

**Comment by kate-porter** [comment, marked as answer]
Reposting the final answer to this question, since it got a bit buried in replies:

> hey @katieporter , at the end we pinned this version and it worked
> 
> "@twilio-paste/core": "4.1.0",
> "@twilio-paste/icons": "3.12.1",
>  "typescript": "3.9.4",
> 
> I will try with the "@types/react": "16.9.35" version whenever I have a chance and I will let you know, thanks a lot!
